src/frontend/js/pages/DashboardCourses/useOrdersEnrollments.tsx

```diff
@@ -0,0 +1,277 @@
+  log('eof', eof.current);
+
+  const fetchEntity = async <T extends Resource>(
+    queryKey: QueryKey,
+    fn: (
+      filters?: ResourcesQuery & { page?: number; page_size?: number },
+    ) => Promise<PaginatedResponse<T>>,
+    filters: any & { page: number; page_size?: number },
+  ): Promise<Maybe<PaginatedResponse<T>>> => {
+    const queryKeyString = queryKey.join('-');
+    // Do not fetch if we already reached the end of the list.
+    if (
+      typeof eof.current[queryKeyString] === 'number' &&
+      filters.page > eof.current[queryKeyString]
+    ) {
+      return;
+    }
+    const QUERY_KEY = ['user', ...queryKey, JSON.stringify(filters)];
+    const data = queryClient.getQueryData<PaginatedResponse<T>>(QUERY_KEY);
+    if (data) {
+      return data;
+    }
+    const res = await fn(filters);
+    queryClient.setQueryData(QUERY_KEY, res);
+
+    // If we reached the end of the list, we set the eof flag to prevent future requests.
+    if (!res.next) {
+      eof.current = { ...eof.current, [queryKeyString]: filters.page };
+      // Eof is cached based, the same way, we cache the fetching data. Otherwise it there would
+      // be request to non existing pages after reload.
+      queryClient.setQueryData(EOF_QUERY_KEY, eof.current);
+    }
+    return res;
+  };
+
+  const fetchEntities = async (pageToLoad: number) => {
+    let totalCount = 0;
+    let entities: Data[] = [];
+    const filters = {
+      page: pageToLoad,
+      page_size: CHUNK_SIZE,
+    };
+
+    const ordersResponse = await fetchEntity<Order>(['orders'], api.user.orders.get, filters);
+    if (ordersResponse) {
+      entities = [
+        ...ordersResponse.results.map((order) => ({ type: DataType.ORDER, item: order })),
+      ];
+      totalCount += ordersResponse.count;
+    }
+
+    const enrollmentsResponse = await fetchEntity<Enrollment>(
+      ['enrollments'],
+      api.user.enrollments.get,
+      { ...filters, was_created_by_order: false },
+    );
+    if (enrollmentsResponse) {
+      entities = [
+        ...entities,
+        ...enrollmentsResponse.results.map((enrollment) => ({
+          type: DataType.ENROLLMENT,
+          item: enrollment,
+        })),
+      ];
+      totalCount += enrollmentsResponse.count;
+    }
+
+    return { entities, totalCount };
+  };
+
+  const fetchEntitiesInStack = async (pageToLoad: number, _stack: Data[]) => {
+    const { entities, totalCount } = await fetchEntities(pageToLoad);
+    // Add missing ones.
+    entities.forEach((entity) => {
+      const index = _stack.findIndex((item) => item.item.id === entity.item.id);
+      if (index === -1) {
+        _stack.push(entity);
+      }
+    });
+    // Sort according to created_on
+    _stack.sort((a, b) => {
+      const aDate = new Date(a.item.created_on);
+      const bDate = new Date(b.item.created_on);
+      return bDate.getTime() - aDate.getTime();
+    });
+
+    return { entities, totalCount };
+  };
+
+  return { fetchEntitiesInStack };
```

Comment:
That would break the algorithm. Here we want to add in place the non existing entities. Where your code returns the non existing ones. Furthermore you remove the sorting which is the key point to guarantee integrity. but maybe I get the point wrong, let's talk about it during daily.